man/cvs.eclass.5
.TH CVS.ECLASS 5 "Nov 2004" "Portage 2.0.51" portage
.SH NAME
cvs.eclass \- provides generic cvs fetching functions
.SH DESCRIPTION
The \fBcvs\fR eclass contains a suite of functions that are used to
create 'live' cvs ebuilds.  These ebuilds are called 'live' because
they will checkout a cvs repository at the time of emerge and compile
the source from the checkout.  This eclass usually generates pretty
unstable ebuilds ... but you can't get anymore bleeding edge than
a live cvs checkout :).

Common usage of this eclass involves setting \fBECVS_SERVER\fR and
\fBECVS_MODULE\fR to grab the relevant cvs sources, and then letting
the eclass define \fBsrc_unpack\fR for you.
.SH VARIABLES
.TP
.B "ECVS_CVS_COMMAND" = \fI"cvs -q -f -z4"\fR
The command to run in order to checkout the module from the cvs
repository.  You should not need to change this.
.TP
.B "ECVS_UP_OPTS" = \fI"-dP"\fR
Options to pass to \fBECVS_CVS_COMMAND\fR when running an update.  
You should not have to modify this.
.TP
.B "ECVS_CO_OPTS" = \fI""\fR
Options to pass to \fBECVS_CVS_COMMAND\fR when running a checkout.  
You should not have to modify this.
.TP
.B "ECVS_TOP_DIR" = \fI"${DISTDIR}/cvs-src"\fR
Where the cvs modules will be stored and then accessed.  You
should not have to modify this.
.TP
.B "ECVS_SERVER" = \fI"offline"\fR
Set this to the server that you will be checking the sources out of.  
If you set this to 'offline' then the eclass will assume the module
is already checked out and in \fBECVS_TOP_DIR\fR.  You almost always
will set this variable.
.TP
.B "ECVS_AUTH" = \fI"pserver"\fR
The authentication method to use to checkout the sources.  Please
note that the only method currently supported is 'pserver'.  You
should not need to modify this variable.
.TP
.B "ECVS_USER" = \fI"anonymous"\fR
The user to log into the server with.
.TP
.B "ECVS_PASS" = \fI""\fR
The password to log into the server with.
.TP
.B "ECVS_MODULE" = \fI""\fR
The module to checkout from the cvs server.  Please note that you
*must* set this variable.
.TP
.B "ECVS_BRANCH" = \fI"HEAD"\fR
The branch to checkout the sources from.  The common targets are
HEAD (current 'stable' cvs code) and SPLIT (current 'unstable' cvs
code).
.SH FUNCTIONS
.TP
.B cvs_fetch
This function will setup \fBECVS_TOP_DIR\fR and any other pre-checkout
steps that may need to be taken.  Then it will login into the server
and finally checkout the sources from cvs.  You usually do not call
this function yourself, but rather let it be handled by \fBcvs_src_unpack\fR.
.TP
.B cvs_src_unpack
This function will determine where the cvs files need to be saved and
then where they need to be placed after the checkout.  The final result
is that you will have a copy of \fBECVS_MODULE\fR in \fB${WORKDIR}\fR.  Thus
you will usually set \fB${S}\fR to be \fB${WORKDIR}/${ECVS_MODULE}\fR.
.SH REPORTING BUGS
Please report bugs via http://bugs.gentoo.org/
.SH SEE ALSO
.BR ebuild (5)
.SH FILES
.BR /usr/portage/eclass/cvs.eclass
.SH AUTHORS
Mike Frysinger <vapier@gentoo.org>
.SH CVS HEADER
$Id: cvs.eclass.5 1912 2005-08-25 03:54:42Z ferringb $


portage/sync/cvs.py
# cvs.py; provides cvs sync capabilities, encapsulates the necessary cvs binary calls
# Copyright 2004 Gentoo Foundation
# Distributed under the terms of the GNU General Public License v2
#$Id: cvs.py 1911 2005-08-25 03:44:21Z ferringb $

import os, stat
from portage.spawn import spawn, spawn_bash, CommandNotFound
#import sync
from portage.const import CVS_BIN

import sync.syncexceptions
class CVSIOError(sync.syncexceptions.SyncException):
	def __init__(self,errmsg,command):
		self.errmsg=errmsg
		self.command=command
	def __str__(self):
		return "cvs error: command %s, %s" % (self.command, self.errmsg)

class CvsHost:
	def __init__(self,host_uri,cvs_binary=CVS_BIN):
		if not os.access(cvs_binary, os.X_OK):
			raise CommandNotFound(cvs_binary)
		self.__binary=cvs_binary
		#parse the bugger.
		#new format.
		#cvs://[CVS_RSH binary:]user@host:cvs_root:module
		#example
		#cvs://ssh:ferringb@dev.gentoo.org:/var/cvsroot:gentoo-x86
		#old format
		#cvs://user@host:cvsroot
		#implicit gentoo-x86 module, and ext w/ ssh.
		#here we go. :/

		if host_uri.count(":") >= 2:
			self.__init_new_host_uri(host_uri)
		else:
			self.__init_deprecated_uri(host_uri)

	def __init_new_host_uri(self,host):
		#cvs://ssh:ferringb@dev.gentoo.org:/var/cvsroot:gentoo-x86
		s=host.split(":")
		if len(s) == 4:
			self.__ext=s.pop(0)
			s[0] = ":ext:" + s[0]
		else:
			self.__ext=None
		self.__cvsroot=s[0]+":"+s[1]
		self.__cvsmodule=s[2]

	def __init_deprecated_uri(self,host):
		self.__ext="ssh"
		self.__cvsmodule="gentoo-x86"
		self.__cvsroot=host

	def sync(self,local_path,verbosity=1,compress=False):
		while local_path[-1] == "/":
			local_path = local_path[:-1]
		if compress:
			c_arg='-z9'
		else:
			c_arg=''

		env={}
		if self.__ext:
			env = {"CVS_RSH":self.__ext}
		
		l=len(self.__cvsmodule)
		if not os.path.exists(local_path):
			newdir=os.path.basename(local_path)
			basedir=local_path[:-len(newdir)]
			if os.path.exists(basedir+"/"+self.__cvsmodule):
				raise Exception("unable to checkout to %s, module directory %s exists already" % \
					(basedir, self.__cvsmodule))
			elif os.path.exists(basedir+"/CVS"):
				raise Exception("unable to checkout to %s, a CVS directory exists w/in already" % basedir)
			command="cd '%s' ; %s %s -d %s co -P  %s" % \
				(basedir, self.__binary, c_arg, self.__cvsroot, self.__cvsmodule)

			ret=spawn_bash(command,env=env,opt_name="cvs co")
			if ret:
				raise CVSIOError("failed checkout",command)
			if newdir != self.__cvsmodule:
				ret = spawn(('mv','%s/%s' % (basedir,self.__cvsmodule),local_path))
				if ret:
					raise Exception("failed moving %s/%s to %s" % (basedir,self.__cvsmodule,local_path))
		elif stat.S_ISDIR(os.stat(local_path).st_mode):

			command="cd '%s'; %s %s -d %s up" % (local_path, self.__binary, c_arg,self.__cvsroot)
			ret = spawn_bash(command, env=env,opt_name="cvs up")
			if ret:
				raise CVSIOError("failed updated", command)
		else:
			raise Exception("%s exists, and is not a directory.  rectify please" % local_path)
		return True

